readme.md
# Matter Plugin For Logseq

This is the official [Matter](https://hq.getmatter.com) plugin for
[Logseq](https://logseq.com/). Sync all of your highlights and notes directly
into your Logseq graph.

![](assets/demo.gif)

---

## Usage
* Install the Matter plugin from Logseq's marketplace
* Enable the plugin
* Open the Matter settings modal and connect to your Matter account
  * Open the Matter application on your phone
  * Go to your Profile > Connected Accounts > Logseq
  * Press "Scan QR Code" and scan the QR code in the Matter settings modal of Logseq
* Configure the plugin how you would like
* Press "Sync Now" to start an initial sync!

CHANGELOG.md
## [1.0.2](https://github.com/getmatterapp/logseq-matter/compare/v1.0.1...v1.0.2) (2023-10-06)


### Bug Fixes

* 1.0.2 ([f26ede9](https://github.com/getmatterapp/logseq-matter/commit/f26ede94a42625b49c02981d4ad8d7438114f7e6))

## [1.0.1](https://github.com/getmatterapp/logseq-matter/compare/v1.0.0...v1.0.1) (2022-10-31)


### Bug Fixes

* 1.0.1 ([a1744bb](https://github.com/getmatterapp/logseq-matter/commit/a1744bb1805cee02a305d4eb81ab560765341ccd))

# 1.0.0 (2022-10-28)


### Features

* v1.0.0 ([258e11c](https://github.com/getmatterapp/logseq-matter/commit/258e11c230f3aef5e8f66dd631686235d0f53932))

## [1.0.0](2022-10-31)


src/api.ts
export const CLIENT_TYPE = 'integration'
export const MATTER_API_VERSION = 'v11'
export const MATTER_API_DOMAIN = 'api.getmatter.app'
export const MATTER_API_HOST = `https://${MATTER_API_DOMAIN}/api/${MATTER_API_VERSION}`
export const ENDPOINTS = {
  QR_LOGIN_TRIGGER: `${MATTER_API_HOST}/qr_login/trigger/`,
  QR_LOGIN_EXCHANGE: `${MATTER_API_HOST}/qr_login/exchange/`,
  REFRESH_TOKEN_EXCHANGE: `${MATTER_API_HOST}/token/refresh/`,
  HIGHLIGHTS_FEED: `${MATTER_API_HOST}/library_items/highlights_feed/`,
}

export interface Annotation {
  created_date: string
  note: string | null
  text: string
  word_start: number
  word_end: number
}

export interface Author {
  any_name: string | null
  domain: string | null
}

export interface ContentNote {
  note: string
}

export interface Publisher {
  any_name: string | null
  domain: string | null
}

export interface Tag {
  created_date: string
  name: string
}
export interface LibraryEntry {
  library_state: number
}

export interface Content {
  author: Author
  library: LibraryEntry | null
  publisher: Publisher
  my_annotations: Annotation[]
  my_note: ContentNote
  publication_date: string
  tags: Tag[]
  title: string
  url: string
}

export interface FeedEntry {
  annotations: Annotation[]
  content: Content
  feed_context: null
  id: string
}

export interface FeedResponse {
  feed: FeedEntry[]
  id: string
  next: string | null
  previous: string | null
}

export interface QRLoginExchangeResponse {
  access_token?: string | null
  refresh_token?: string | null
}

class RequestError extends Error {
  response: Response

  public constructor(response: Response, message?: string) {
    super(message)
    this.response = response
  }
}

export async function authedRequest(
  accessToken: string,
  url: string,
  fetchArgs: RequestInit = {},
) {
  const headers = new Headers()
  headers.set('Authorization', `Bearer ${accessToken}`)
  headers.set('Content-Type', 'application/json')

  const response = await fetch(url, {
    ...fetchArgs,
    headers,
  })

  if (!response.ok) {
    throw new RequestError(response, 'Matter authenticated request failed')
  }

  return await response.json()
}


src/auth.ts
import { CLIENT_TYPE, ENDPOINTS, QRLoginExchangeResponse } from './api'
import { sleep } from './utils'

export async function fetchQRSessionToken() {
  try {
    const headers = new Headers()
    headers.set('Content-Type', 'application/json')
    const triggerResponse = await fetch(ENDPOINTS.QR_LOGIN_TRIGGER, {
      method: 'POST',
      body: JSON.stringify({ client_type: CLIENT_TYPE }),
      headers,
    })
    return (await triggerResponse.json()).session_token
  } catch (error) {
    return null
  }
}

export async function pollQRLoginExchange(sessionToken: string) {
  let attempts = 0
  while (attempts < 600) {
    try {
      const loginSession = await qrLoginExchange(sessionToken)
      if (loginSession?.access_token) {
        return {
          access_token: loginSession.access_token,
          refresh_token: loginSession.refresh_token,
        }
      }
    } finally {
      attempts++
      await sleep(1000)
    }
  }
}

async function qrLoginExchange(
  sessionToken: string,
): Promise<QRLoginExchangeResponse> {
  const headers = new Headers()
  headers.set('Content-Type', 'application/json')
  const response = await fetch(ENDPOINTS.QR_LOGIN_EXCHANGE, {
    method: 'POST',
    body: JSON.stringify({
      session_token: sessionToken,
    }),
    headers,
  })
  return response.json()
}


src/utils.ts
import { LSPluginUserEvents } from '@logseq/libs/dist/LSPlugin.user'
import React from 'react'

let _visible = logseq.isMainUIVisible

function subscribeLogseqEvent<T extends LSPluginUserEvents>(
  eventName: T,
  handler: (...args: any) => void,
) {
  logseq.on(eventName, handler)
  return () => {
    logseq.off(eventName, handler)
  }
}

const subscribeToUIVisible = (onChange: () => void) =>
  subscribeLogseqEvent('ui:visible:changed', ({ visible }) => {
    _visible = visible
    onChange()
  })

export const useAppVisible = () => {
  return React.useSyncExternalStore(subscribeToUIVisible, () => _visible)
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export function randomInt(range: number) {
  const sign = Math.round(Math.random()) * 2 - 1
  const abs = Math.floor(Math.random() * range)
  return sign * abs
}

export function diffInMinutes(a: Date, b: Date): number {
  const diffMs = (a as any) - (b as any)
  const diffS = diffMs / 1000
  return diffS / 60
}


src/sync.ts
import { BlockEntity, PageEntity } from '@logseq/libs/dist/LSPlugin.user'
import {
  Annotation,
  authedRequest,
  ENDPOINTS,
  FeedEntry,
  FeedResponse,
  Tag,
} from './api'
import {
  syncJitterRange,
  syncStaleMinutes,
  syncFrequencies,
  syncCooldownMinutes,
  maxNumWrites,
} from './constants'
import { diffInMinutes, randomInt } from './utils'
import { format } from 'date-fns'

export const intervalSync = () => {
  const syncJitter = randomInt(syncJitterRange)

  // If a sync has been abandoned, restart the process
  if (isSyncStale(syncJitter) && logseq.settings!.matterIsSyncing) {
    sync()
    return
  }

  const now = new Date()
  const lastSync = getLastSync()
  const syncFrequencyKey = logseq.settings!.matterSyncFrequency

  let syncFrequency = syncFrequencies[syncFrequencyKey]
  if (syncFrequency < 0) {
    return
  }
  syncFrequency += syncJitter

  let should = false
  if (lastSync) {
    if (syncFrequency > 0 && diffInMinutes(now, lastSync) >= syncFrequency) {
      should = true
    }
  } else {
    should = true
  }

  // If a sync hasn't happened in some time, start a new one
  if (should && !logseq.settings!.matterIsSyncing) {
    sync()
  }
}

export const sync = async () => {
  // The graph used on first sync should be the graph targeted for every
  // subsequent sync
  const currentGraph = await logseq.App.getCurrentGraph()
  if (!logseq.settings?.graph) {
    logseq.updateSettings({
      graph: currentGraph,
    })
  } else if (logseq.settings.graph.url !== currentGraph!.url) {
    return
  }

  logseq.UI.showMsg('Syncing with Matter...', 'success', {
    timeout: 3000,
  })
  syncHeartbeat()
  setIsSyncing(true)

  try {
    const complete = await pageAnnotations()
    setLastSync(new Date())
    if (complete) {
      setIsSyncing(false)
      logseq.UI.showMsg('Finished syncing with Matter', 'success', {
        timeout: 3000,
      })
    } else {
      setTimeout(sync, 60 * syncCooldownMinutes * 1000)
    }
  } catch (error) {
    console.error(error)
  }
}

const pageAnnotations = async (): Promise<boolean> => {
  let url: string | null = ENDPOINTS.HIGHLIGHTS_FEED
  let feedEntries: FeedEntry[] = []

  // Load all feed items new to old.
  while (url !== null) {
    syncHeartbeat()
    const response: FeedResponse = await _authedRequest(url)
    feedEntries = feedEntries.concat(response.feed)
    url = response.next
  }

  // Reverse the feed items so that chronological ordering is preserved.
  feedEntries = feedEntries.reverse()

  let writeCount = 0
  for (const feedEntry of feedEntries) {
    const written = await handleFeedEntry(feedEntry)

    if (written) {
      writeCount += 1
    }

    if (writeCount >= maxNumWrites) {
      return false
    }
  }

  return true
}

const handleFeedEntry = async (feedEntry: FeedEntry): Promise<boolean> => {
  const pageTitle = feedEntry.content.title.trim()

  let searchTitle = pageTitle
  while (searchTitle.startsWith('#')) {
    searchTitle = pageTitle.slice(1).trim()
  }

  const page = await logseq.Editor.getPage(searchTitle)
  if (page) {
    const lastSync = getLastSync()
    let annotations = feedEntry.content.my_annotations

    if (lastSync) {
      annotations = annotations.filter(
        (a) => new Date(a.created_date) > lastSync,
      )
    }

    const pageTree = await logseq.Editor.getPageBlocksTree(page.name)
    annotations = annotations.filter(
      (a) => !annotationAppearsInPage(a, pageTree),
    )

    if (annotations.length) {
      await appendAnnotationsToPage(page, annotations)
      return true
    }

    return false
  } else {
    const page = await logseq.Editor.createPage(
      pageTitle,
      {},
      { redirect: false, createFirstBlock: false },
    )
    if (page) {
      await renderPage(page, feedEntry)
      return true
    }
    return false
  }
}

const renderPage = async (page: PageEntity, feedEntry: FeedEntry) => {
  const meta: string[] = [
    `source:: [${feedEntry.content.title}](${feedEntry.content.url})`,
  ]

  if (feedEntry.content.author) {
    if (feedEntry.content.author.any_name) {
      meta.push(`author:: [[${feedEntry.content.author.any_name}]]`)
    } else if (feedEntry.content.author.domain) {
      meta.push(`author:: [[${feedEntry.content.author.domain}]]`)
    }
  }

  if (feedEntry.content.publisher) {
    if (feedEntry.content.publisher.any_name) {
      meta.push(`publisher:: [[${feedEntry.content.publisher.any_name}]]`)
    } else {
      meta.push(`publisher:: [[${feedEntry.content.publisher.domain}]]`)
    }
  }

  if (feedEntry.content.tags) {
    meta.push(`tags:: ${renderTags(feedEntry.content.tags)}`)
  }

  if (feedEntry.content.my_note && feedEntry.content.my_note.note) {
    meta.push(`note:: ${feedEntry.content.my_note.note}`)
  }

  await logseq.Editor.appendBlockInPage(page.uuid, meta.join('\n'))
  await appendAnnotationsToPage(page, feedEntry.content.my_annotations)
}

const appendAnnotationsToPage = async (
  page: PageEntity,
  annotations: Annotation[],
) => {
  if (!annotations.length) {
    return
  }
  annotations = annotations.sort((a, b) => a.word_start - b.word_start)
  const userConfig = await logseq.App.getUserConfigs()
  const todayStr = format(new Date(), userConfig.preferredDateFormat)

  let highlightsHeader: string
  if (userConfig.enabledJournals) {
    highlightsHeader = `#### [[Highlights]] synced from [[Matter]] on [[${todayStr}]]`
  } else {
    highlightsHeader = `#### [[Highlights]] synced from [[Matter]] on [[${todayStr}]]`
  }

  let highlightsSection: BlockEntity | null | undefined = (
    await logseq.Editor.getPageBlocksTree(page.name)
  ).find((block) => block.content === highlightsHeader)

  if (!highlightsSection) {
    highlightsSection = await logseq.Editor.appendBlockInPage(
      page.uuid,
      highlightsHeader,
    )
  }

  for (const annotation of annotations) {
    const properties: { [key: string]: string } = {}
    if (annotation.note) {
      properties.note = annotation.note
    }

    await logseq.Editor.appendBlockInPage(
      highlightsSection?.uuid || page.uuid,
      annotation.text,
      {
        properties,
      },
    )
  }
}

const annotationAppearsInPage = (
  annotation: Annotation,
  pageTree: BlockEntity[],
): boolean => {
  while (pageTree.length) {
    const block = pageTree.pop()

    if (block?.content === annotation.text) {
      return true
    }

    if (block && block.children) {
      pageTree = pageTree.concat(block.children as BlockEntity[])
    }
  }

  return false
}

const renderTags = (tags: Tag[]): string => {
  const tagStrs = tags.map((tag) => {
    return `#[[${tag.name}]]`
  })

  if (tagStrs.length) {
    return ` ${tagStrs.join(' ')}`
  }

  return ''
}

const _authedRequest = async (url: string) => {
  const accessToken = logseq.settings!.matterAccessToken
  try {
    return await authedRequest(accessToken, url)
  } catch (e) {
    await refreshTokenExchange()
    return await authedRequest(accessToken, url)
  }
}

const refreshTokenExchange = async () => {
  const refreshToken = logseq.settings!.matterRefreshToken

  const headers = new Headers()
  headers.set('Content-Type', 'application/json')
  const response = await fetch(ENDPOINTS.REFRESH_TOKEN_EXCHANGE, {
    method: 'POST',
    headers,
    body: JSON.stringify({ refresh_token: refreshToken }),
  })
  const payload = await response.json()
  logseq.updateSettings({
    matterAccessToken: payload.access_token,
    matterRefreshToken: payload.refresh_token,
  })
}

const setIsSyncing = (val: boolean) => {
  logseq.updateSettings({
    matterIsSyncing: val,
  })
}

const setLastSync = (val: Date) => {
  logseq.updateSettings({
    matterLastSync: val.toISOString(),
  })
}

const syncHeartbeat = () => {
  logseq.updateSettings({
    matterSyncHeartbeat: new Date().toISOString(),
  })
}

const isSyncStale = (syncJitter: number) => {
  const lastHeartbeat = logseq.settings!.matterSyncHeartbeat
  if (lastHeartbeat) {
    return (
      diffInMinutes(new Date(), new Date(lastHeartbeat)) >
      syncStaleMinutes + syncJitter
    )
  }
  return true
}

const getLastSync = (): Date | null => {
  const timestamp = logseq.settings!.matterLastSync
  if (timestamp) {
    return new Date(timestamp)
  }
  return null
}
